File: java/shift2DGridBM.java
class Solution {
        private static final int REMEDY = 1000;   
    private static final int numOfBits = 11;
    public List<List<Integer>> shiftGrid(int[][] grid, int k) {
        int R = grid.length, C = grid[0].length, total = R * C;
        k %= total;
        makeAllNoLessThan0(grid); 
        saveNewValsByHigherBits(grid, k);
        recoverToOriginalVals(grid);
     // return Arrays.stream(grid)
     //              .map(row -> IntStream.of(row).boxed().collect(Collectors.toList()))
     //              .collect(Collectors.toList());
        LinkedList<List<Integer>> ans = new LinkedList<>();
        for (int[] row : grid) {
            ans.offer(new ArrayList<>());
            for (int cell : row) {
                ans.peekLast().add(cell);
            }
        }
        return ans;
    }
    private void makeAllNoLessThan0(int[][] grid) {
        for (int[] row : grid) {
            for (int i = 0; i < row.length; ++i) {
                row[i] += REMEDY;
            }
        }
    }
    private void saveNewValsByHigherBits(int[][] grid, int k) {
        for (int r = 0, R = grid.length, C = grid[0].length, mask = ((1 << numOfBits) - 1); r < R; ++r) {
            for (int c = 0; c < C; ++c) {
                int target = (r * C + c + k) % (R * C), i = target / C, j = target % C;
                grid[i][j] |= (grid[r][c] & mask) << numOfBits;
            }
        }
    }
    private void recoverToOriginalVals(int[][] grid) {
        for (int[] row : grid) {
            for (int c = 0; c < row.length; ++c) {
                row[c] >>= numOfBits;
                row[c] -= REMEDY;
            }
        }
    }
}

File: Python/shift2DGridBM.py
from typing import List


class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:

        def makeAllNoLessThan0(remedy: int) -> None:
            for row in grid:
                for c, _ in enumerate(row):
                    row[c] += remedy

        def saveNewValsByHigherBits(numOfBits: int) -> None:
            mask = 2 ** numOfBits - 1
            for r in range(R):
                for c in range(C):
                    target = (r * C + c + k) % total
                    i, j = target // C, target % C 
                    grid[i][j] |= (grid[r][c] & mask) << numOfBits

        def recoverToOriginalVals(remedy: int, numOfBits: int) -> None:
            for row in grid:
                for c, _ in enumerate(row):
                    row[c] >>= numOfBits
                    row[c] -= remedy

        R, C = map(len, (grid, grid[0]))
        total = R * C
        k %= total

        makeAllNoLessThan0(1000)
        saveNewValsByHigherBits(11)
        recoverToOriginalVals(1000, 11)

        return grid